caproto/threading/client.py

```diff
@@ -1494,9 +1494,9 @@ def __init__(self, pv):
         self._callback_lock = threading.RLock()
 
     def add_callback(self, func):
-        # TODO thread safety
-        cb_id = self._callback_id
-        self._callback_id += 1
+        with self._callback_lock:
```

Comment:
:+1: threadsafety - perhaps combine into one block, moving out all things that don't need locking?

Reply:
There's no way around this ordering (nothing to move outside the lock) but I guess it's probably better to acquire once and do a little extra work in the locked state that to thrash the lock twice in quick succession.